### DISCUSSION TITLE
Nuxt2 to Nuxt3 migration: how to deal with runtime config ?

### DISCUSSION THREAD

**Discussion by JBustin** [nuxt/nuxt#26491] [Questions]
Hello,

I'm migrating a huge project from nuxt2 to nuxt3, and I get a blocking point to migrate the runtime config.

My nuxt2 runtime config is working like this:
- **based on .env**
- variable names in code are generally different of environment variable names
- **populated during runtime**, because we build an image for different environments (which seems to me to be banal...)
- accept some transformations - for example cast or decode operations - because we chose to expose a configuration without any required additional operation to be effective

Example:
```js
publicRuntimeConfig: {
  has_bar: process.env.NAMESPACE_BAR_ENABLE == 'true',
  foo_data: decode(process.env.NAMESPACE_FOO_ENCODED, {}),
  bar_content: process.env.NAMESPACE_BAR || '',
}
```

I'm trying to figure out how to migrate to nuxt3... but I'm still failing.

## With nuxt3 runtime config

If I want to populate my config at runtime, I need to use the prefix syntax.
BUT, 
- **Break change**: I have to rename my variables to use the PREFIX mechanism to be able to load the variable on **runtime**
- **Blocker**: I cant use transformations.

```js
runtimeConfig: {
  nitro: {
    prefix: 'NAMESPACE_',
  },
  public: {
    barEnable: process.env.NAMESPACE_BAR_ENABLE == 'true',
    fooEncoded: decode(process.env.NAMESPACE_FOO_ENCODED, {}),
    bar: '',
  }
}
```

`barEnable` and `fooEncoded` are not set cause of the function or cast usage.

## With nuxt3 app config

Great, especially the HMR that can avoid the long long and painful server rebuild and restart and development when we change a variable in .env.
BUT, 
- **Blocker**: I can't use .env without tricks.
- And I'm not sure to be able to use it with option api.

## Finally

It seems to be painful to migrate this important file (~100 variables).
Why runtime config comes with so much constraints ?
What can I do ? 
What do you suggest ?
A custom implementation ?

I'm using a stackblitz to test different solutions: https://stackblitz.com/edit/nuxt3-welcome-4xa3ge?file=nuxt.config.ts,.env

**Comment by xjccc** [comment]
[serialization](https://nuxt.com/docs/guide/going-further/runtime-config#serialization)

[Nuxt's runtimeConfig](https://www.youtube.com/watch?v=_FYV5WfiWvs)

I hope these helped.

**Reply by JBustin** [reply]
Thank you for taking the time to respond.

_Instead of passing non-serializable objects or functions into your application from your nuxt.config, you can place this code in a Nuxt or Nitro plugin or middleware._

I guess, it means, to move my nuxt2 config to nuxt3, I need to:
- update the runtime config to use the PREFIX, and simple assignations (no cast and no transformation)
- create a plugin named, I don't know, `myAppConfig`, that takes the runtime config and apply all the cast and transformations.
- replace in the code all the references to `ctx.$config` by `ctx.$myAppConfig` (because directly extend $config in a plugin has very bad side effects -  I tried)

Another thing that doesn't make migration easier

I have an additional feature to implement: to extend my current config with a remote config hosted in a private http registry.
I'm trying to use app.config instead of runtimeConfig, because there is a method `updateAppConfig`. I think that `myAppConfig` will include this logic too. 
I'm at the extreme limit of a custom implementation ^^

Nothing easier ?

**Reply by xjccc** [reply]
I'm not nuxt2 to nuxt3
I can't give you some simple way to migrate😂
Maybe someone could give you some suggestion.

**Comment by JBustin** [comment]
I finally choose to use `app.config` to replace the `runtimeConfig`, with some tricks:
- Before starting server, I generate the `app.config.ts` file by converting my previous `runtimeConfig` with the loading of the .env
- a plugin `0.app-config.ts` is in charge to inject `$appConfig` for option API usage
- I replace all the references of `$config` by `$appConfig` in my code

I'm a quite disappointed to implement this kind of stuff, which should be managed by the framework from my point of view.
